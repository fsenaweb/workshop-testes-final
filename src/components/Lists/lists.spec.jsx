import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import user from "@testing-library/user-event";

import Lists from ".";

describe("<Lists />", () => {
  it("Renderiza o componente", () => {
    render(<Lists />);

    // Verifica se o form, input e button estão renderizados
    expect(screen.getByTestId("form-add")).toBeInTheDocument();
    expect(screen.getByTestId("input")).toBeInTheDocument();
    expect(screen.getByText(/salvar/i)).toBeInTheDocument();
  });

  it("Ao digitar menos de 3 caracteres o botão ficará desabilitado", async () => {
    render(<Lists />);

    const button = screen.getByText(/salvar/i);
    const input = screen.getByTestId("input");

    expect(button).toBeDisabled();

    const text = "12";
    user.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
    });

    expect(button).toBeDisabled();
  });

  it("Ao digitar 3 caracteres ou mais, o botão ficará habilitado", async () => {
    render(<Lists />);
    const button = screen.getByText(/salvar/i);
    const input = screen.getByTestId("input");

    expect(button).toBeDisabled();

    const text = "123";
    user.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
    });

    expect(button).toBeEnabled();
  });

  it("Deve exibir um novo item ao clicar no botão salvar", async () => {
    render(<Lists />);

    const input = screen.getByTestId("input");
    const submit = screen.getByText("Salvar");

    fireEvent.change(input, { target: { value: "Casa Nova" } });
    fireEvent.click(submit);

    const element = screen.getByText("Casa Nova");

    expect(element).toBeInTheDocument();
  });

  it("Deve contar a quantidade de itens após cadastros", () => {
    render(<Lists />);

    const input = screen.getByTestId("input");
    const submit = screen.getByText("Salvar");

    // Cadastrando uma palavra
    fireEvent.change(input, { target: { value: "Palavra 1" } });
    fireEvent.click(submit);

    fireEvent.change(input, { target: { value: "Palavra 2" } });
    fireEvent.click(submit);

    const listItems = screen.getAllByTestId("item");

    expect(listItems.length).toBe(2);
  });

  it("Não deve adicionar um item duplicado", () => {
    render(<Lists />);

    const input = screen.getByTestId("input");
    const submit = screen.getByText("Salvar");

    fireEvent.change(input, { target: { value: "Roupa" } });
    fireEvent.click(submit);

    fireEvent.click(submit);

    fireEvent.change(input, { target: { value: "Roupa" } });
    fireEvent.click(submit);

    const element = screen.getAllByText("Roupa");
    expect(element.length).toBe(1);
  });

  it("Deleta as palavras ao clicar no botão delete", () => {
    const { container } = render(<Lists />);

    const input = screen.getByTestId("input");
    const submit = screen.getByText("Salvar");

    fireEvent.change(input, { target: { value: "Casa Nova" } });
    fireEvent.click(submit);

    const deleteButton = screen.getByTestId("Casa Nova-btn-delete");
    fireEvent.click(deleteButton);

    const deletedElement = screen.queryByText("Casa Nova");
    expect(deletedElement).not.toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
