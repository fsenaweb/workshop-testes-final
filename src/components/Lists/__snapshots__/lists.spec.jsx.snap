// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Lists /> Deleta as palavras ao clicar no bot√£o delete 1`] = `
<div>
  <div
    class="text-body-secondary py-3 my-5 border-top"
  >
    <h3
      class="mb-3 px-3"
    >
      Lista de Tarefas
    </h3>
    <form
      class="container mb-4"
      data-testid="form-add"
    >
      <div
        class="row"
      >
        <div
          class="col-10"
        >
          <input
            class="form-control"
            data-testid="input"
            type="text"
            value=""
          />
        </div>
        <div
          class="col-2 d-grid"
        >
          <button
            class="btn btn-primary btn"
            disabled=""
            type="submit"
          >
            Salvar
          </button>
        </div>
      </div>
    </form>
    <div
      class="container"
    >
      <ul
        class="list-group"
        data-testid="ul-itens"
      />
    </div>
  </div>
</div>
`;
